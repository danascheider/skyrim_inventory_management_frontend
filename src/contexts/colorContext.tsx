import { createContext, ReactElement } from 'react'
import { YELLOW, type ColorScheme } from '../utils/colorSchemes'

const ColorContext = createContext<ColorScheme>(YELLOW)

interface ProviderProps {
  colorScheme: ColorScheme
  children: ReactElement
}

const ColorProvider = ({ colorScheme, children }: ProviderProps) => (
  <ColorContext.Provider value={colorScheme}>{children}</ColorContext.Provider>
)

export { ColorProvider, ColorContext }
