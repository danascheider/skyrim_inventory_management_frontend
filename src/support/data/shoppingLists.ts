import {
  type ResponseShoppingList as ShoppingList,
  type ResponseShoppingListItem as ListItem,
} from '../../types/apiData'

/**
 *
 * Shopping lists are related to games in that each shopping list belongs to a
 * particular game. Shopping lists in this file have a `game_id` corresponding
 * to one of the games in the `allGames` array exported from /src/support/data/games.ts.
 * The empty shopping lists array could hypothetically belong to any of the games.
 *
 */

const emptyListItems: ListItem[] = []

export const emptyShoppingLists: ShoppingList[] = []

export const allShoppingLists: ShoppingList[] = [
  {
    id: 1,
    game_id: 32,
    aggregate_list_id: null,
    aggregate: true,
    title: 'All Items',
    list_items: emptyListItems,
    created_at: new Date('2023-01-02T03:54:02'),
    updated_at: new Date('2023-01-02T03:54:02'),
  },
  {
    id: 2,
    game_id: 32,
    aggregate_list_id: 1,
    aggregate: false,
    title: 'My Shopping List 1',
    list_items: emptyListItems,
    created_at: new Date('2023-01-02T03:54:02'),
    updated_at: new Date('2023-01-02T03:54:02'),
  },
  {
    id: 3,
    game_id: 77,
    aggregate_list_id: null,
    aggregate: true,
    title: 'All Items',
    list_items: emptyListItems,
    created_at: new Date('2023-02-12T15:17:33'),
    updated_at: new Date('2023-02-12T15:17:33'),
  },
  {
    id: 4,
    game_id: 77,
    aggregate_list_id: 3,
    aggregate: false,
    title: 'Honeyside',
    list_items: emptyListItems,
    created_at: new Date('2023-02-21T11:13:27'),
    updated_at: new Date('2023-02-21T11:13:27'),
  },
  {
    id: 5,
    game_id: 77,
    aggregate_list_id: 3,
    aggregate: false,
    title: 'Breezehome',
    list_items: emptyListItems,
    created_at: new Date('2023-02-12T15:17:33'),
    updated_at: new Date('2023-02-12T15:17:33'),
  },
]

export const shoppingListsForGame = (gameId: number) =>
  allShoppingLists.filter(({ game_id }) => game_id === gameId)
