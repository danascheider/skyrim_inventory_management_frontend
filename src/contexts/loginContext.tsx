import { createContext, useState, useEffect, useCallback } from 'react'
import { type User } from 'firebase/auth'
import { useAuthState } from 'react-firebase-hooks/auth'
import { auth, signOutWithGoogle } from '../firebase'
import { ProviderProps } from '../types/contexts'

export interface LoginContextType {
  authLoading: boolean
  token: string | null
  withTokenRefresh: (fn: (idToken: string) => void) => void
  user?: User | null
}

export const LoginContext = createContext<LoginContextType>({
  user: null,
  token: null,
  authLoading: true,
  withTokenRefresh: () => {},
})

export const LoginProvider = ({ children }: ProviderProps) => {
  const [user, authLoading, authError] = useAuthState(auth)
  const [token, setToken] = useState<string | null>(null)

  const refreshToken = useCallback(() => {
    if (user) user.getIdToken(true).then((idToken) => setToken(idToken))
  }, [user])

  const withTokenRefresh = (fn: Function) => {
    if (user) {
      user.getIdToken(true).then((idToken) => {
        setToken(idToken)
        fn(idToken)
      })
    }
  }

  const value = {
    user,
    token,
    withTokenRefresh,
    authLoading,
  }

  useEffect(() => {
    if (authError) signOutWithGoogle()

    refreshToken()
  }, [user, authError])

  return <LoginContext.Provider value={value}>{children}</LoginContext.Provider>
}
