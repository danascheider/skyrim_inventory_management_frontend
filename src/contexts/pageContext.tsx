import { createContext, useState, useEffect, useRef } from 'react'
import { type ApiCalls, type Resource, type HttpVerb } from '../types/apiCalls'
import { ProviderProps } from '../types/contexts'
import { type FlashProps, type ModalProps } from '../types/pageContext'

interface PageContextType {
  flashProps: FlashProps
  setFlashProps: (props: FlashProps) => void
  modalProps: ModalProps
  setModalProps: (props: ModalProps) => void
  apiCallsInProgress: ApiCalls
  addApiCall: (key: Resource, value: HttpVerb) => void
  removeApiCall: (key: Resource, value: HttpVerb) => void
}

const defaultFlashProps: FlashProps = {
  type: 'info',
  message: '',
  hidden: true,
}

const defaultModalProps: ModalProps = {
  hidden: true,
  children: <></>,
}

const defaultApiCallStatus = {
  games: [],
  shoppingLists: [],
  shoppingListItems: [],
}

const PageContext = createContext<PageContextType>({
  flashProps: defaultFlashProps,
  modalProps: defaultModalProps,
  apiCallsInProgress: defaultApiCallStatus,
  setFlashProps: () => {},
  setModalProps: () => {},
  addApiCall: () => {},
  removeApiCall: () => {},
})

const PageProvider = ({ children }: ProviderProps) => {
  const [flashProps, setFlashProps] = useState<FlashProps>(defaultFlashProps)
  const [modalProps, setModalProps] = useState<ModalProps>(defaultModalProps)
  const [apiCallsInProgress, SetApiCallsInProgress] =
    useState<ApiCalls>(defaultApiCallStatus)

  const flashVisibleSince = useRef(0)

  const addApiCall = (key: Resource, value: HttpVerb) => {
    SetApiCallsInProgress({
      ...apiCallsInProgress,
      [key]: [...apiCallsInProgress[key], value],
    })
  }

  const removeApiCall = (key: Resource, value: HttpVerb) => {
    SetApiCallsInProgress({
      ...apiCallsInProgress,
      [key]: apiCallsInProgress[key].filter((verb) => verb !== value),
    })
  }

  const value = {
    flashProps,
    modalProps,
    apiCallsInProgress,
    setFlashProps,
    setModalProps,
    addApiCall,
    removeApiCall,
  }

  useEffect(() => {
    if (flashProps.hidden === true) return

    flashVisibleSince.current = Number(new Date())

    setTimeout(() => {
      if (Number(new Date()) - flashVisibleSince.current >= 4000) {
        setFlashProps({ ...flashProps, hidden: true })
        flashVisibleSince.current = 0
      }
    }, 4000)
  }, [flashProps])

  return <PageContext.Provider value={value}>{children}</PageContext.Provider>
}

export { PageContext, PageProvider }
