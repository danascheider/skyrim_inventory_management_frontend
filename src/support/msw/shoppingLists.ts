import { rest } from 'msw'
import { allGames } from '../data/games'
import { allShoppingLists } from '../data/shoppingLists'
import { shoppingListsForGame } from '../data/shoppingLists'
import { newShoppingList, newShoppingListWithAggregate } from './helpers/data'

const BASE_URI = 'http://localhost:3000'
const gameIds = allGames.map(({ id }) => id)
const listIds = allShoppingLists.map(({ id }) => id)

/**
 *
 * POST /games/:game_id/shopping_lists
 *
 */

// Handles both 201 and 404 responses
export const postShoppingLists = rest.post(
  `${BASE_URI}/games/:gameId/shopping_lists`,
  async (req, res, ctx) => {
    const gameId: number = Number(req.params.gameId)

    if (gameIds.indexOf(gameId) < 0) return res(ctx.status(404))

    const attributes = await req.json()

    const responseBody = shoppingListsForGame(gameId).length
      ? newShoppingList(attributes, gameId)
      : newShoppingListWithAggregate(attributes, gameId)

    return res(ctx.status(201), ctx.json(responseBody))
  }
)

// Returns the same validation errors regardless of request body
// submitted
export const postShoppingListsUnprocessable = rest.post(
  `${BASE_URI}/games/:gameId/shopping_lists`,
  (_req, res, ctx) => {
    return res(
      ctx.status(422),
      ctx.json({
        errors: [
          'Title must be unique per game',
          "Title can only contain alphanumeric characters, spaces, commas (,), hyphens (-), and apostrophes (')",
        ],
      })
    )
  }
)

export const postShoppingListsServerError = rest.post(
  `${BASE_URI}/games/:gameId/shopping_lists`,
  (_req, res, ctx) => {
    return res(
      ctx.status(500),
      ctx.json({
        errors: ['Something went horribly wrong'],
      })
    )
  }
)

/**
 *
 * GET /games/:game_id/shopping_lists
 *
 */

// Covers both success and 404 cases
export const getShoppingLists = rest.get(
  `${BASE_URI}/games/:gameId/shopping_lists`,
  (req, res, ctx) => {
    const gameId: number = Number(req.params.gameId)
    if (gameIds.indexOf(gameId) < 0) return res(ctx.status(404))

    return res(
      ctx.status(200),
      ctx.json(shoppingListsForGame(Number(req.params.gameId)))
    )
  }
)

/**
 *
 * PATCH /shopping_lists/:id
 *
 */

// Covers both success and 404 cases
export const patchShoppingList = rest.patch(
  `${BASE_URI}/shopping_lists/:id`,
  async (req, res, ctx) => {
    const listId: number = Number(req.params.id)
    if (listIds.indexOf(listId) < 0) return res(ctx.status(404))

    const list = allShoppingLists.find(({ id }) => id === listId)
    const { title } = await req.json()

    return res(ctx.status(200), ctx.json({ ...list, title }))
  }
)

// Returns the same validation errors regardless of request
// body submitted
export const patchShoppingListUnprocessable = rest.patch(
  `${BASE_URI}/shopping_lists/:id`,
  (_req, res, ctx) => {
    return res(
      ctx.status(422),
      ctx.json({
        errors: [
          'Title must be unique per game',
          "Title can only contain alphanumeric characters, spaces, commas (,), hyphens (-), and apostrophes (')",
        ],
      })
    )
  }
)

export const patchShoppingListServerError = rest.patch(
  `${BASE_URI}/shopping_lists/:id`,
  (_req, res, ctx) => {
    return res(
      ctx.status(500),
      ctx.json({
        errors: ['Something went horribly wrong'],
      })
    )
  }
)

/**
 *
 * DELETE /shopping_lists/:id
 *
 */

// Covers both success and 404 cases
export const deleteShoppingList = rest.delete(
  `${BASE_URI}/shopping_lists/:listId`,
  (req, res, ctx) => {
    const listId = Number(req.params.listId)
    const list = allShoppingLists.find(({ id }) => id === listId)

    if (!list) return res(ctx.status(404))

    const lists = shoppingListsForGame(list.game_id)

    let json
    if (lists.length === 2) {
      json = {
        deleted: lists.map(({ id }) => id),
      }
    } else {
      json = {
        deleted: [list.id],
        aggregate: lists[0],
      }
    }

    return res(ctx.status(200), ctx.json(json))
  }
)

export const deleteShoppingListServerError = rest.delete(
  `${BASE_URI}/shopping_lists/:listId`,
  (_req, res, ctx) => {
    return res(
      ctx.status(500),
      ctx.json({
        errors: ['Something went horribly wrong'],
      })
    )
  }
)
