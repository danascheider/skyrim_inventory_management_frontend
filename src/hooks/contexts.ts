import { useContext } from 'react'
import { ColorContext } from '../contexts/colorContext'
import { LoginContext } from '../contexts/loginContext'

const useCustomContext = <T>(cxt: React.Context<T>, msg: string) => {
  const context = useContext(cxt)

  if (!context) throw new Error(msg)

  return context
}

export const useColorScheme = () =>
  useCustomContext(
    ColorContext,
    'useColorScheme must be used within a ColorProvider'
  )

export const useGoogleLogin = () =>
  useCustomContext(
    LoginContext,
    'useGoogleLogin must be used within a LoginProvider'
  )