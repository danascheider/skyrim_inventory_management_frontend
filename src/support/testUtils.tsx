import { type ReactElement } from 'react'
import { BrowserRouter } from 'react-router-dom'
import { JSDOM } from 'jsdom'
import { render } from '@testing-library/react'
import { LoginContext } from '../contexts/loginContext'
import { User } from 'firebase/auth'
import testProfileImg from './testProfileImg.png'

export const BASE_APP_URI = 'http://localhost:5173'

declare global {
  namespace NodeJS {
    interface Global {
      document: Document
      window: Window
    }
  }
}

const setDom = () => {
  const dom = new JSDOM('<!doctype html><html><body></body></html>', {
    url: 'http://localhost:5173',
  })

  global.window = dom.window as unknown as Window & typeof globalThis
  global.document = dom.window.document
}

/*
 *
 * Global Test Data
 *
 */

export const requireLogin = () => {
  /* noop */
}

export const testUser = {
  uid: 'edna',
  displayName: 'Edna St. Vincent Millay',
  email: 'edna@gmail.com',
  photoURL: testProfileImg,
  emailVerified: true,
  getIdtoken: () =>
    new Promise<string>((resolve, _reject) => resolve('xxxxxxx')),
} as unknown as User

/*
 *
 * Test Renderers
 *
 */

const renderWithJSDOM = (ui: ReactElement) => {
  setDom()

  return render(ui)
}

export const renderWithRouter = (ui: ReactElement) =>
  renderWithJSDOM(<BrowserRouter>{ui}</BrowserRouter>)

export const renderAuthenticated = (ui: ReactElement, authLoading = false) =>
  renderWithRouter(
    <LoginContext.Provider
      value={{
        user: testUser,
        token: 'xxxxxxx',
        requireLogin,
        authLoading,
      }}
    >
      {ui}
    </LoginContext.Provider>
  )

export const renderUnauthenticated = (ui: ReactElement) =>
  renderWithRouter(
    <LoginContext.Provider
      value={{
        user: null,
        token: null,
        authLoading: false,
        requireLogin,
      }}
    >
      {ui}
    </LoginContext.Provider>
  )

export const renderAuthLoading = (ui: ReactElement) =>
  renderWithRouter(
    <LoginContext.Provider
      value={{ user: null, token: null, authLoading: true, requireLogin }}
    >
      {ui}
    </LoginContext.Provider>
  )
