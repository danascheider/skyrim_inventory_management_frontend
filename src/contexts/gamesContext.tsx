import { createContext, useCallback, useEffect, useState } from 'react'
import { signOutWithGoogle } from '../firebase'
import { type Game } from '../types/games'
import { type ProviderProps } from '../types/contexts'
import { useGoogleLogin, usePageContext } from '../hooks/contexts'
import { ApiError } from '../types/errors'
import { LOADING, DONE, ERROR, type LoadingState } from '../utils/loadingStates'
import { getGames, deleteGame } from '../utils/simApi'

// TODO: Add default values for context provider as follows:
//       const createGame = (game: Game) => {
//                                            postGames(game: Game, token: '')
//                                              .then((status: number, json: string) => /* noop */)
//                                          }
//       This will be useful in Storybook.

const NOT_FOUND_MESSAGE =
  "Oops! We couldn't find the game you're looking for. Please refresh and try again."
const UNEXPECTED_ERROR_MESSAGE =
  "Oops! Something unexpected went wrong. We're sorry! Please try again later."

export interface GamesContextType {
  games: Game[]
  gamesLoadingState: LoadingState
  // createGame: (game: Game) => void
  // updateGame: (gameId: number, attrs: Game) => void
  destroyGame: (gameId: number) => void
}

export const GamesContext = createContext<GamesContextType>({
  games: [],
  gamesLoadingState: LOADING,
  destroyGame: () => {},
})

export const GamesProvider = ({ children }: ProviderProps) => {
  const { user, token, authLoading, requireLogin } = useGoogleLogin()
  const [gamesLoadingState, setGamesLoadingState] = useState(LOADING)
  const [games, setGames] = useState<Game[]>([])
  const { setFlashProps } = usePageContext()

  /*
   *
   * Retrieve all the current user's games from the API
   *
   */

  const fetchGames = useCallback(() => {
    if (user && token) {
      getGames(token)
        .then(({ json }) => {
          if (Array.isArray(json)) {
            setGames(json)
            setGamesLoadingState(DONE)
          }
        })
        .catch((e: ApiError) => {
          if (import.meta.env.DEV)
            console.error(`Error ${e.name}: ${e.message}`)

          if (e.code === 401) {
            signOutWithGoogle()
          } else {
            setGames([])
            setGamesLoadingState(ERROR)
            setFlashProps({
              type: 'error',
              message: e.message,
              hidden: false,
            })
          }
        })
    }
  }, [user, token])

  /*
   *
   * Destroy the requested game and update the `games` array
   *
   */

  const destroyGame = useCallback(
    (gameId: number) => {
      if (user && token) {
        deleteGame(gameId, token)
          .then(({ status }) => {
            if (status === 204) {
              const newGames = games.filter(({ id }) => id !== gameId)
              setGames(newGames)
              setFlashProps({
                hidden: false,
                type: 'success',
                message: 'Success! Your game has been deleted.',
              })
            }
          })
          .catch((e: ApiError) => {
            if (import.meta.env.DEV)
              console.log(`Error ${e.name}: ${e.message}`)

            const message =
              e.code === 404 ? NOT_FOUND_MESSAGE : UNEXPECTED_ERROR_MESSAGE

            setFlashProps({
              hidden: false,
              type: 'error',
              message,
            })
          })
      }
    },
    [user, token, games]
  )

  const value = {
    games,
    gamesLoadingState,
    destroyGame,
  }

  useEffect(() => {
    requireLogin()
  }, [requireLogin])

  useEffect(() => {
    if (authLoading) return

    fetchGames()
  }, [authLoading, fetchGames])

  return <GamesContext.Provider value={value}>{children}</GamesContext.Provider>
}
