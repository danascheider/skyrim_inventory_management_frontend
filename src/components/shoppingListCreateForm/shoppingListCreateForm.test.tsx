import { type ReactElement } from 'react'
import { describe, test, expect, beforeAll, beforeEach, afterAll } from 'vitest'
import { act, fireEvent } from '@testing-library/react'
import { setupServer } from 'msw/node'
import { renderAuthenticated } from '../../support/testUtils'
import { getShoppingListsSuccess } from '../../support/msw/shoppingLists'
import {
  gamesContextValue,
  gamesContextValueError,
  gamesContextValueLoading,
  shoppingListsContextValue,
  shoppingListsContextValueError,
  shoppingListsContextValueLoading,
} from '../../support/data/contextValues'
import { PageProvider } from '../../contexts/pageContext'
import { GamesContext } from '../../contexts/gamesContext'
import {
  ShoppingListsContext,
  ShoppingListsProvider,
} from '../../contexts/shoppingListsContext'
import ShoppingListCreateForm from './shoppingListCreateForm'

const renderWithContexts = (ui: ReactElement, url?: string) => {
  return renderAuthenticated(
    <PageProvider>
      <GamesContext.Provider value={gamesContextValue}>
        <ShoppingListsProvider>{ui}</ShoppingListsProvider>
      </GamesContext.Provider>
    </PageProvider>,
    url
  )
}

describe('ShoppingListCreateForm', () => {
  describe('displaying the form', () => {
    const mockServer = setupServer(getShoppingListsSuccess)

    beforeAll(() => mockServer.listen())
    beforeEach(() => mockServer.resetHandlers())
    afterAll(() => mockServer.close())

    test('has the correct form fields', () => {
      const wrapper = renderWithContexts(<ShoppingListCreateForm />)

      expect(wrapper).toBeTruthy()

      expect(wrapper.getByPlaceholderText('Title')).toBeTruthy()
      expect(wrapper.getByText('Create')).toBeTruthy()
    })

    test('matches snapshot', () => {
      const wrapper = renderWithContexts(<ShoppingListCreateForm />)

      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('disabling behaviour', () => {
    test('is disabled when the shopping lists are loading', () => {
      const wrapper = renderAuthenticated(
        <PageProvider>
          <GamesContext.Provider value={gamesContextValue}>
            <ShoppingListsContext.Provider
              value={shoppingListsContextValueLoading}
            >
              <ShoppingListCreateForm />
            </ShoppingListsContext.Provider>
          </GamesContext.Provider>
        </PageProvider>
      )

      expect(
        wrapper
          .getByPlaceholderText('Title')
          .attributes.getNamedItem('disabled')
      ).toBeTruthy()
      expect(
        wrapper.getByText('Create').attributes.getNamedItem('disabled')
      ).toBeTruthy()
    })

    test('is disabled when there is a shopping list loading error', () => {
      const wrapper = renderAuthenticated(
        <PageProvider>
          <GamesContext.Provider value={gamesContextValue}>
            <ShoppingListsContext.Provider
              value={shoppingListsContextValueError}
            >
              <ShoppingListCreateForm />
            </ShoppingListsContext.Provider>
          </GamesContext.Provider>
        </PageProvider>
      )

      expect(
        wrapper
          .getByPlaceholderText('Title')
          .attributes.getNamedItem('disabled')
      ).toBeTruthy()
      expect(
        wrapper.getByText('Create').attributes.getNamedItem('disabled')
      ).toBeTruthy()
    })

    test('is disabled when games are loading', () => {
      const wrapper = renderAuthenticated(
        <PageProvider>
          <GamesContext.Provider value={gamesContextValueLoading}>
            <ShoppingListsContext.Provider value={shoppingListsContextValue}>
              <ShoppingListCreateForm />
            </ShoppingListsContext.Provider>
          </GamesContext.Provider>
        </PageProvider>
      )

      expect(
        wrapper
          .getByPlaceholderText('Title')
          .attributes.getNamedItem('disabled')
      ).toBeTruthy()
      expect(
        wrapper.getByText('Create').attributes.getNamedItem('disabled')
      ).toBeTruthy()
    })

    test('is disabled when there is a game loading error', () => {
      const wrapper = renderAuthenticated(
        <PageProvider>
          <GamesContext.Provider value={gamesContextValueError}>
            <ShoppingListsContext.Provider value={shoppingListsContextValue}>
              <ShoppingListCreateForm />
            </ShoppingListsContext.Provider>
          </GamesContext.Provider>
        </PageProvider>
      )

      expect(
        wrapper
          .getByPlaceholderText('Title')
          .attributes.getNamedItem('disabled')
      ).toBeTruthy()
      expect(
        wrapper.getByText('Create').attributes.getNamedItem('disabled')
      ).toBeTruthy()
    })

    test('is enabled when both games and shopping lists have loaded', () => {
      const wrapper = renderAuthenticated(
        <PageProvider>
          <GamesContext.Provider value={gamesContextValue}>
            <ShoppingListsContext.Provider value={shoppingListsContextValue}>
              <ShoppingListCreateForm />
            </ShoppingListsContext.Provider>
          </GamesContext.Provider>
        </PageProvider>
      )

      expect(
        wrapper
          .getByPlaceholderText('Title')
          .attributes.getNamedItem('disabled')
      ).toBeFalsy()
      expect(
        wrapper.getByText('Create').attributes.getNamedItem('disabled')
      ).toBeFalsy()
    })
  })

  describe('submitting the form', () => {
    describe('when the form is enabled', () => {
      test('calls the createShoppingList function', () => {
        const createShoppingList = vitest.fn()
        const contextValue = {
          ...shoppingListsContextValue,
          createShoppingList,
        }

        const wrapper = renderAuthenticated(
          <PageProvider>
            <GamesContext.Provider value={gamesContextValue}>
              <ShoppingListsContext.Provider value={contextValue}>
                <ShoppingListCreateForm />
              </ShoppingListsContext.Provider>
            </GamesContext.Provider>
          </PageProvider>
        )

        const input = wrapper.getByPlaceholderText('Title')
        const button = wrapper.getByText('Create')

        act(() => {
          fireEvent.change(input, { target: { value: 'New Shopping List' } })
          fireEvent.click(button)
        })

        expect(createShoppingList).toHaveBeenCalledWith(
          { title: 'New Shopping List' },
          expect.any(Function)
        )
      })
    })

    describe('when the form is disabled', () => {
      test("doesn't call the createShoppingList function", () => {
        const createShoppingList = vitest.fn()
        const contextValue = {
          ...shoppingListsContextValueLoading,
          createShoppingList,
        }

        const wrapper = renderAuthenticated(
          <PageProvider>
            <GamesContext.Provider value={gamesContextValue}>
              <ShoppingListsContext.Provider value={contextValue}>
                <ShoppingListCreateForm />
              </ShoppingListsContext.Provider>
            </GamesContext.Provider>
          </PageProvider>
        )

        const button = wrapper.getByText('Create')

        act(() => {
          fireEvent.click(button)
        })

        expect(createShoppingList).not.toHaveBeenCalled()
      })
    })
  })
})
