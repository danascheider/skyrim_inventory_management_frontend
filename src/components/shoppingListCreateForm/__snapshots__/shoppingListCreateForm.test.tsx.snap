// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ShoppingListCreateForm > displaying the form > matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="_root_1118ad"
        style="--button-color: #4e8fb7; --button-text-color: #fff; --button-border-color: #1b5c84; --button-hover-color: #3881ae;"
      >
        <fieldset
          class="_fieldset_1118ad"
        >
          <input
            aria-label="Title"
            class="_input_1118ad"
            disabled=""
            name="title"
            pattern="\\\\s*[A-Za-z0-9 \\\\-',]*\\\\s*"
            placeholder="Title"
            title="Title can only contain alphanumeric characters, spaces, commas, hyphens, and apostrophes"
            type="text"
          />
          <button
            class="_button_1118ad"
            disabled=""
            type="submit"
          >
            Create
          </button>
        </fieldset>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="_root_1118ad"
      style="--button-color: #4e8fb7; --button-text-color: #fff; --button-border-color: #1b5c84; --button-hover-color: #3881ae;"
    >
      <fieldset
        class="_fieldset_1118ad"
      >
        <input
          aria-label="Title"
          class="_input_1118ad"
          disabled=""
          name="title"
          pattern="\\\\s*[A-Za-z0-9 \\\\-',]*\\\\s*"
          placeholder="Title"
          title="Title can only contain alphanumeric characters, spaces, commas, hyphens, and apostrophes"
          type="text"
        />
        <button
          class="_button_1118ad"
          disabled=""
          type="submit"
        >
          Create
        </button>
      </fieldset>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
